Include parameters in interactive provider diff

acmcelwee opened the pull request:
We often use the interactive provider when we're dealing w/ more sensitive stacks, like dynamo tables, vpc, etc, but the current state of interactive only prompts for approval if there are structural or resource changes to that cloudformation detects, and the prompt omits any information about parameters that are changing. To me, parameters are a critical aspect of the changeset, so they deserve more direct attention in the interactive provider's prompt for approval.

This leans on some of the existing formatting of param changes that are used in the `diff` command, but it limits the parameters shown to ones that are actually changed, removed, or newly added. Any parameters that remain unchanged are not included in the prompt output.

Here's a snippet from an interactive build that adds a param `AdamTest` and changes the `VersionLabel` param value.

```
[2017-06-05T23:24:30] dptDwapi changes:
Parameters Added: AdamTest
Parameters Removed: ParamThatWasRemoved
Parameters Modified: VersionLabel, Foo
Replacements:
- Modify Task (AWS::ECS::TaskDefinition)
Changes:
- Modify CPUUtilizationHigh (AWS::CloudWatch::Alarm)
- Modify CPUUtilizationLow (AWS::CloudWatch::Alarm)
- Modify ScaleInPolicy (AWS::ApplicationAutoScaling::ScalingPolicy)
- Modify ScaleOutPolicy (AWS::ApplicationAutoScaling::ScalingPolicy)
- Modify ScalingTarget (AWS::ApplicationAutoScaling::ScalableTarget)
- Modify Service (AWS::ECS::Service)
- Modify ServiceRole (AWS::IAM::Role)
Execute the above changes? [y/n/v]
```

In verbose mode, the pull params are output exactly the same as the `diff` command does today.

Example:

```
[2017-06-22T18:40:01] Full changeset:

--- Old Parameters
+++ New Parameters
******************
 UnchangedParam = unchangedValue
+AdamTest = testValue
-ParamThatWasRemoved
-VersionLabel = 0.0.1
+VersionLabel = 0.0.2

- ResourceChange:
    Action: Modify
    Details:
    - CausingEntity: ScaleInPolicy
...
```

phobologic commented:
This looks cool - we tend to just rely on the fact that the objects will change if the parameters change, but I think this is a good enhancement.  A couple of comments (without diving into the code yet):

- I'm a little concerned about the use of `-` and `+` to indicate changes, since later in the text we use `-` to indicate a bullet.  Might lead to quite a bit of confusion.
- It might be better to just list what parameters have changed in the normal mode, and then display the actual changes if someone hits `v` to get more verbose output. We've tried to keep the normal mode somewhat concise.

Let me know what you think - and thanks for this!  I'll try to dive into the code soon!

acmcelwee commented:
@phobologic 100% w/ ya on the confusion of `-` and `+` in the context of the list output of the resource changes below. So, maybe just a list of variables that are either: new, removed, or changed in the initial prompt, and then the `-`/`+` output for verbose?

phobologic commented:
@acmcelwee I think that's a good method - in the non-verbose mode, maybe it just says:

```
Added Parameters: Param1, Param2
Removed Parameters: Param3
Modified Parameters: Param4
```

In the verbose mode those could be sections with the actual values. Maybe use the format like we use in the diff command? https://github.com/remind101/stacker/blob/master/stacker/actions/diff.py#L67

phobologic commented:
Hey @acmcelwee - just wanted to ping you on this - I'm probably going to do a release today, but we'll probably do another soon.

acmcelwee commented:
@phobologic updated

acmcelwee commented:
@phobologic after we discussed the revised representation in non-verbose mode, I added a `DictValue` wrapper to capture the param diffs output. Since the usage of those dict diffs expanded beyond just joining and printing them out, it seemed more clean to carry that info around in a more intelligent, first-class type. It cleaned up the filtering necessary to partition changes into the `Added`, `Removed`, and `Modified` buckets to properly display a non-verbose summary.

As a side note, I squashed my changes down into a single commit and force-pushed, because you guys tend to merge (vs squash merge). I always feel bad when I have 3 commits of the form "fix line length" because I forgot to check the flake8 stuff 😄 . Those commit messages just cloud up the commit history if they're merged in, as is, so I just wanted this to merge a bit cleaner. Do you guys just never squash merge as a rule?

phobologic commented on stacker/providers/aws/interactive.py:
```diff
@@ -205,23 +255,27 @@ def __init__(self, region, replacements_only=False, *args, **kwargs):
 
-        if changes:
-            output_summary(fqn, action, changes,
+        if changes or params_diff:
```
I'm not sure this should be checked - if the parameters change, but it doesn't actually cause any changes, do you think it's still worth interrupting the flow/getting approval?

phobologic replied:
Maybe there should be a note if there's a parameter change, but no changeset changes  - just to let people know that it's not an error?  Any ideas on that?

acmcelwee replied:
So, I actually just tried quite a few different ways to trigger that scenario, but since stacker discards unused parameters, I'm not sure that the scenario even exists to worry about.

phobologic commented:
Thanks @acmcelwee!
